### PR TITLE
Prepare simulator to target wasm32

### PR DESCRIPTION
**Pull request by fabioarnold**
We can't have long-running functions in a JavaScript environment. That's why I had to split the main function in simulator.zig.

TODO:

- [x] Split `main` into `init` and `tick`
- [x] Setup logging to output to JS `console.log`
- [x] Don't crash during initialization (fix memory bugs)

### CODE REVIEW

File: src/config.zig [hunk 1]
```diff
@@ -228,7 +235,9 @@ pub const journal_size_prepares = journal_slot_count * message_size_max;
 
 comptime {
     // vsr.parse_address assumes that config.address/config.port are valid.
-    _ = std.net.Address.parseIp4(address, 0) catch unreachable;
+    if (!is_wasm) {
```
Comment (sentientwaffle) [inline src/config.zig]: What error does this give when compiled for wasm?

File: src/simulator.zig [hunk 2]
```diff
@@ -25,199 +25,200 @@ const log_health = std.log.scoped(.health);
 /// You can fine tune your log levels even further (debug/info/notice/warn/err/crit/alert/emerg):
```
Comment (sentientwaffle) [inline src/simulator.zig]: Since these changes are pretty comprehensive, in the interest of not blocking you (and avoiding git conflicts with several other PRs):

Please copy all of your changes to (e.g.) `simulator_frontend.zig`, and leave `simulator.zig` untouched in your PR. Once the vopr-hub is merged, I will integrate your changes into `simulator.zig` (e.g. the separate `tick()` function), and you can switch back.

File: src/test/storage.zig [hunk 3]
```diff
@@ -116,14 +116,14 @@ pub const Storage = struct {
         assert(options.write_latency_mean >= options.write_latency_min);
```
Comment (sentientwaffle) [inline src/test/storage.zig:116]: @jorangreef Why are Storage's offsets/sizes `u64` instead of `usize`?

File: src/config.zig [hunk 4]
```diff
@@ -1,8 +1,11 @@
 const std = @import("std");
+const builtin = @import("builtin");
 const assert = std.debug.assert;
 
+const is_wasm = builtin.target.isWasm();
+
```
Comment (sentientwaffle) [inline src/config.zig]: Inlining these for clarity (to avoid an extra local variable).

```suggestion
```

File: src/config.zig [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
 /// Whether development or production:
-pub const deployment_environment = .development;
+pub const deployment_environment = if (is_wasm) .simulator_frontend else .development;
```
Comment (sentientwaffle) [inline src/config.zig]:
```suggestion
pub const deployment_environment = if (builtin.target.isWasm()) .simulator_frontend else .development;
```

File: src/config.zig (hunk 1 above)
Comment (sentientwaffle) [inline src/config.zig]:
```suggestion
    if (!builtin.target.isWasm()) {
```

File: src/vsr/replica.zig [hunk 6]
```diff
@@ -1465,7 +1465,7 @@ pub fn Replica(
             // This saves us from going to disk. And we don't need to worry that the WAL's copy
             // of an uncommitted prepare is lost/corrupted.
             if (self.pipeline_prepare_for_op_and_checksum(op, checksum)) |prepare| {
-                log.debug("{}: on_request_prepare: op={} checksum={} reply from pipeline", .{
+                log.debug("{}: on_request_prepare: op={} checksum={any} reply from pipeline", .{
```
Comment (sentientwaffle) [inline src/vsr/replica.zig:1468]: Why are these changed? `checksum` is an integer, which should print just fine with `{}`.

File: src/vsr/replica.zig [hunk 7]
```diff
@@ -4135,7 +4135,7 @@ pub fn Replica(
                 received.* = null;
             }
             assert(count <= self.replica_count);
-            log.debug("{}: reset {} {s} message(s) from view={}", .{
+            log.debug("{}: reset {} {s} message(s) from view={any}", .{
```
Comment (sentientwaffle) [inline src/vsr/replica.zig:4118]: Also an integer.

File: src/config.zig (hunk 1 above)
Comment (fabioarnold) [reply src/config.zig]:
```
/opt/homebrew/Cellar/zig/HEAD-e863292_1/lib/zig/std/os.zig:177:28: error: container 'std.c' has no member called 'sockaddr'
pub const sockaddr = system.sockaddr;
```
`freestanding` is missing sockets I think.

File: src/simulator.zig (hunk 2 above)
Comment (fabioarnold) [reply src/simulator.zig]: Sounds good! 👍🏻

File: src/vsr/replica.zig (hunk 6 above)
Comment (fabioarnold) [reply src/vsr/replica.zig:1468]: `checksum` is an optional. These errors only show with Zig master.

Comment (sentientwaffle) [reply src/vsr/replica.zig:1468]: Ah, ok, thanks!

File: src/vsr/replica.zig (hunk 7 above)
Comment (fabioarnold) [reply src/vsr/replica.zig:4118]: `view` is `?u32`.

File: src/test/storage.zig (hunk 3 above)
Comment (jorangreef) [reply src/test/storage.zig:116]: To be explicit around the size of data file that we support, regardless of architecture.

In other words, `usize` is more for where we explicitly want to use a machine word-sized integer, for example, for performance when writing an encoding/decoding routine that processes data in the data plane using word-sized integers—think of all the tricks that Daniel Lemire uses around machine words.

Whereas, here, for `offsets/sizes`, we're interested in addressing data files larger than 4GiB. i.e. We don't mind using two machine words on a 32-bit architecture.

Where it's not in the data plane, and where there's no specific performance reason to do otherwise, `u64` is generally almost always a better bet than `usize` since it's not going to overflow, and since it's not subject to non-determinism based on architecture.

Where memory is a concern, and where `u32` is the right choice, then that's also better. We save memory on 64-bit architectures, and we're also still more explicit and deterministic than a `usize`.

I know that Zig's std lib almost always just uses `usize` but I prefer that this is reserved only for where the intention is literally to use machine words in the data plane. For the control plane, `u64` (default) or `u32` (memory-intensive) are better deterministic choices. They're also more explicit.

**Comment by jorangreef** [comment]
Closing since we've branched since this, let me know if we should re-open!
